### PR TITLE
PimpleAwareEventDispatcher

### PR DESCRIPTION
**Pull request by Atriedes**
make current event dispatcher aware of service in pimple.

It provide similiar feature like this : http://symfony.com/doc/current/components/event_dispatcher/container_aware_dispatcher.html

this code adapted from : https://github.com/symfony/EventDispatcher/blob/master/ContainerAwareEventDispatcher.php

### CODE REVIEW

**Comment by dosten** [comment]
Why `PimpleAwareEventDispatcher` and not `ContainerAwareEventDispatcher`?

File: src/Silex/PimpleAwareEventDispatcher.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
+<?php
+
+
```
Comment (dosten) [inline src/Silex/PimpleAwareEventDispatcher.php]: Extra line

File: src/Silex/PimpleAwareEventDispatcher.php [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * @param Container $container
+     */
+    public function __construct(Container $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * @param $event_name
```
Comment (dosten) [inline src/Silex/PimpleAwareEventDispatcher.php]: Missing param type

File: src/Silex/PimpleAwareEventDispatcher.php [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+     * @param $callback
```
Comment (dosten) [inline src/Silex/PimpleAwareEventDispatcher.php]: Missing param type

File: src/Silex/PimpleAwareEventDispatcher.php [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        } else {
+            $this->lazyLoad($event_name);
+        }
+
+        return parent::getListeners($event_name);
+    }
+
+    /**
+     * @param $service_id
+     * @param $class
```
Comment (dosten) [inline src/Silex/PimpleAwareEventDispatcher.php]: Same here

File: src/Silex/PimpleAwareEventDispatcher.php [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+     */
+    public function addSubscriberService($service_id, $class)
```
Comment (dosten) [inline src/Silex/PimpleAwareEventDispatcher.php]: You need to type hint the `$class` parameter with `Symfony\Component\EventDispatcher\EventSubscriberInterface`

**Comment by Atriedes** [comment]
@dosten : why not ContainerAwareEventDispatcher because it only implement container which is pimple, as you see in here https://github.com/symfony/EventDispatcher/blob/master/ContainerAwareEventDispatcher.php it also implement specific container. So I decided to use PimpleAware to avoid confusion.

And thanks for code review, will updated soon :)

File: src/Silex/PimpleAwareEventDispatcher.php (hunk 5 above)
Comment (Atriedes) [inline src/Silex/PimpleAwareEventDispatcher.php]: `$class` must be in a string that contain class name in order to perform lazy load for EvenSubscriber.

please refer to this docs : http://symfony.com/doc/current/components/event_dispatcher/container_aware_dispatcher.html#adding-subscriber-services

**Comment by fabpot** [comment]
I'm :-1: on adding so much code to Silex. I'm not even sure if it's worth it.

**Comment by alexantr** [comment]
@fabpot
> I'm :-1: on adding so much code to Silex.

:+1:

**Pull request closed without merging**